src/AutoSchedule.cpp

```diff
@@ -0,0 +1,2312 @@
+
+    void merge_groups(const FusionChoice &choice, const EvalConfig &eval,
+                      Partitioner::Level level);
+
+    EvalConfig evaluate_choice(const FusionChoice &fuse, Partitioner::Level level);
+
+    Group fuse_groups(const Group &g1, const Group &g2);
+
+    GroupAnalysis analyze_group(const Group &g, bool show_analysis);
+
+    map<FStage, map<FStage, DimBounds>> group_loop_bounds();
+    map<FStage, map<string, Box>> group_storage_bounds();
+
+    void group(Partitioner::Level level);
+
+    vector<pair<FusionChoice, EvalConfig>>
+    choose_candidate_fuse(const vector<pair<string, string>> &cand_pairs,
+                          Partitioner::Level level);
+
+    map<string, int64_t> evaluate_reuse(const FStage &stg,
+                                        const set<string> &prods);
+
+    map<string, int64_t>
+            analyze_spatial_locality(const FStage &stg,
+                                     const map<string, Box> &parent_bounds);
+
+    int64_t find_max_access_stride(string var, string func_acc,
+                                   vector<Expr> acc_exprs,
+                                   const Box &buffer_bounds);
+
+    map<string, int> bounds_to_estimates(const DimBounds &bounds);
+
+    string generate_cpu_schedule(const Target &t);
+
+    string generate_group_cpu_schedule(const Group &g, const Target &t,
+                                       const map<FStage, DimBounds> &group_loop_bounds,
+                                       const map<string, Box> &group_storage_bounds);
+
+    DimBounds get_bounds(const FStage &stg);
+
+    DimBounds get_bounds_from_tile_sizes(const FStage &stg,
+                                         const map<string, int> &tile_sizes);
+
+    vector<map<string, int>> generate_tile_configs(const FStage &stg);
+
+    pair<map<string, int>, GroupAnalysis>
+            find_best_tile_config(const Group &g, Partitioner::Level level);
+
+    int64_t estimate_benefit(const GroupAnalysis &nofuse, const GroupAnalysis &fuse,
+                             bool no_redundant_work, bool ensure_parallelism);
+
+    int64_t estimate_benefit(const vector<pair<FusionChoice, EvalConfig>> &choices,
+                             bool no_redundant_work, bool ensure_parallelism);
+
+    void initialize_groups_inline();
+    void initialize_groups_fast_mem();
+
+    pair<int64_t, int64_t> get_pipeline_cost();
+
+    void disp_pipeline_costs();
+    void disp_pipeline_bounds();
+    void disp_pipeline_graph();
+    void disp_grouping();
+};
+
+Partitioner::Partitioner(map<string, Box> &_pipeline_bounds,
+                         MachineParams &_arch_params,
+                         DependenceAnalysis &_dep_analy,
+                         RegionCosts &_costs,
+                         const vector<Function> &_outputs,
+                         bool _gpu_schedule):
+    pipeline_bounds(_pipeline_bounds), arch_params(_arch_params),
+    dep_analy(_dep_analy), costs(_costs), outputs(_outputs),
```

Comment:
"analy" is an unfortunate abbreviation. May I suggest spelling out "analysis" instead.